runtime/libia2/init.c

```diff
@@ -13,7 +15,45 @@
 extern __thread void *ia2_stackptr_0[PAGE_SIZE / sizeof(void *)]
     __attribute__((aligned(4096)));
 
-uintptr_t ia2_stack_addrs[MAX_PKEYS] IA2_SHARED_DATA = {0};
+__thread uintptr_t ia2_stack_addrs[MAX_PKEYS] IA2_SHARED_DATA = {0};
+
+typedef struct {
+  pid_t tid;
+  uintptr_t (*stack_addrs)[MAX_PKEYS];
+} IA2ThreadData;
+
+IA2ThreadData *ia2_thread_data_ptr IA2_SHARED_DATA = NULL;
+size_t ia2_thread_data_len IA2_SHARED_DATA = 0;
+size_t ia2_thread_data_cap IA2_SHARED_DATA = 0;
+
+/// Get the `IA2ThreadData` for the current thread,
+/// adding it if it's not already in the array,
+/// and reallocating as necessary.
+IA2ThreadData *ia2_thread_data_get_current_thread(void) {
+  const pid_t tid = gettid();
+  for (size_t i = 0; i < ia2_thread_data_len; i++) {
+    if (ia2_thread_data_ptr[i].tid == tid) {
+      return &ia2_thread_data_ptr[i];
+    }
+  }
+  if (ia2_thread_data_cap == 0) {
+    // Start capacity at 16 instead of 1, 2, 4, 8, 16.
+    ia2_thread_data_cap = 16;
+  }
+  if (ia2_thread_data_len == ia2_thread_data_cap) {
+    ia2_thread_data_cap = ia2_thread_data_cap == 0 ? 16 : ia2_thread_data_cap * 2;
+    ia2_thread_data_ptr = realloc(ia2_thread_data_ptr, ia2_thread_data_cap);
```

Comment:
This should be a `shared_realloc`, right, just like the globals are `IA2_SHARED_DATA`?